Custom DLQ Destination Resolver (#976)
spring-cloud/spring-cloud-stream-binder-kafka: docs/src/main/asciidoc/kafka-streams.adoc

You can set the topic name where the DLQ messages are published as below.

You can provide an implementation for `DlqDestinationResolver` which is a functional interface.
`DlqDestinationResolver` takes `ConsumerRecord` and the exception as inputs and then allows to specify a topic name as the output.
By gaining access to the Kafka `ConsumerRecord`, the header records can be introspected in the implementation of the `BiFunction`.

Here is an example of providing an implementation for `DlqDestinationResolver`.

[source]
----
@Bean
public DlqDestinationResolver dlqDestinationResolver() {
    return (rec, ex) -> {
        if (rec.topic().equals("word1")) {
            return "topic1-dlq";
        }
        else {
            return "topic2-dlq";
        }
    };
}
----

One important thing to keep in mind when providing an implementation for `DlqDestinationResolver` is that the provisioner in the binder will not auto create topics for the application.

> This is because there is no way for the binder to infer all the DLQ topics the implementation is creating or their names.

Reviewer: ```suggestion
This is because there is no way for the binder to infer the names of all the DLQ topics the implementation might send to.
```

Author: @garyrussell Do you recommend the message channel binder changes for this in this PR or as a separate one?

Reviewer: I personally would do it in the same PR, but it's up to you.
